File: pnpm/crates/versioning/src/intents.rs
use std::{
    fs,
    io::ErrorKind,
    path::{Path, PathBuf},
};

use derive_more::Display;
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use crate::{error::VersioningError, human_id::random_human_id, settings::ReleaseBumpType};

/// The directory holding change-intent files, shared with changesets.
pub const CHANGES_DIR: &str = ".changeset";

/// A bump type as recorded in an intent file: a release bump or the additive
/// `none` decline ("this change needs no release").
#[derive(Debug, Display, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum IntentBumpType {
    #[display("none")]
    None,
    #[display("patch")]
    Patch,
    #[display("minor")]
    Minor,
    #[display("major")]
    Major,
}

impl IntentBumpType {
    /// The release bump this entry demands; `None` for a decline.
    #[must_use]
    pub fn release(self) -> Option<ReleaseBumpType> {
        match self {
            IntentBumpType::None => None,
            IntentBumpType::Patch => Some(ReleaseBumpType::Patch),
            IntentBumpType::Minor => Some(ReleaseBumpType::Minor),
            IntentBumpType::Major => Some(ReleaseBumpType::Major),
        }
    }

    fn parse(value: &str) -> Option<IntentBumpType> {
        match value {
            "none" => Some(IntentBumpType::None),
            "patch" => Some(IntentBumpType::Patch),
            "minor" => Some(IntentBumpType::Minor),
            "major" => Some(IntentBumpType::Major),
            _ => None,
        }
    }
}

/// One `.changeset/*.md` file: which packages a change affects, the bump type
/// for each, and the summary that becomes the changelog entry.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChangeIntent {
    pub id: String,
    pub file_path: PathBuf,
    pub releases: IndexMap<String, IntentBumpType>,
    pub summary: String,
}

pub fn parse_change_intent(
    content: &str,
    id: &str,
    file_path: &Path,
) -> Result<ChangeIntent, VersioningError> {
    let lines: Vec<&str> = content
        .trim_start_matches('\u{FEFF}')
        .split('\n')
        .map(|line| line.strip_suffix('\r').unwrap_or(line))
        .collect();
    let closing_index = if lines.first().is_some_and(|line| line.trim() == "---") {
        lines.iter().skip(1).position(|line| line.trim() == "---").map(|index| index + 1)
    } else {
        None
    };
    let Some(closing_index) = closing_index else {
        return Err(VersioningError::NoFrontmatter { file_path: file_path.to_path_buf() });
    };

    let frontmatter_text = lines[1..closing_index].join("\n");
    let frontmatter: IndexMap<String, String> = if frontmatter_text.trim().is_empty() {
        IndexMap::new()
    } else {
        serde_saphyr::from_str(&frontmatter_text).map_err(|err| {
            VersioningError::InvalidFrontmatter {
                file_path: file_path.to_path_buf(),
                message: err.to_string(),
            }
        })?
    };

    let mut releases = IndexMap::new();
    for (pkg_name, bump_type) in frontmatter {
        let Some(parsed) = IntentBumpType::parse(&bump_type) else {
            return Err(VersioningError::InvalidBumpType {
                file_path: file_path.to_path_buf(),
                pkg_name,
                bump_type,
            });
        };
        releases.insert(pkg_name, parsed);
    }

    Ok(ChangeIntent {
        id: id.to_string(),
        file_path: file_path.to_path_buf(),
        releases,
        summary: lines[closing_index + 1..].join("\n").trim().to_string(),
    })
}

pub fn read_change_intents(workspace_dir: &Path) -> Result<Vec<ChangeIntent>, VersioningError> {
    let changes_dir = workspace_dir.join(CHANGES_DIR);
    let entries = match fs::read_dir(&changes_dir) {
        Ok(entries) => entries,
        Err(err) if err.kind() == ErrorKind::NotFound => return Ok(Vec::new()),
        Err(source) => return Err(VersioningError::Read { path: changes_dir, source }),
    };

    let mut file_names = Vec::new();
    for entry in entries {
        let entry =
            entry.map_err(|source| VersioningError::Read { path: changes_dir.clone(), source })?;
        let name = entry.file_name().to_string_lossy().into_owned();
        if name.ends_with(".md") && !name.eq_ignore_ascii_case("readme.md") {
            file_names.push(name);
        }
    }
    file_names.sort();

    file_names
        .into_iter()
        .map(|file_name| {
            let file_path = changes_dir.join(&file_name);
            let content = fs::read_to_string(&file_path)
                .map_err(|source| VersioningError::Read { path: file_path.clone(), source })?;
            let id = file_name.trim_end_matches(".md");
            parse_change_intent(&content, id, &file_path)
        })
        .collect()
}

pub fn write_change_intent(
    workspace_dir: &Path,
    releases: &IndexMap<String, IntentBumpType>,
    summary: &str,
) -> Result<String, VersioningError> {
    let changes_dir = workspace_dir.join(CHANGES_DIR);
    fs::create_dir_all(&changes_dir)
        .map_err(|source| VersioningError::Write { path: changes_dir.clone(), source })?;

    let mut id = random_human_id();
    while changes_dir.join(format!("{id}.md")).exists() {
        id = random_human_id();
    }

    let frontmatter_lines: Vec<String> = releases
        .iter()
        .map(|(pkg_name, bump_type)| {
            format!("{}: {bump_type}", serde_json::to_string(pkg_name).expect("serialize string"))
        })
        .collect();
    let content = format!("---\n{}\n---\n\n{}\n", frontmatter_lines.join("\n"), summary.trim());
    let file_path = changes_dir.join(format!("{id}.md"));
    fs::write(&file_path, content)
        .map_err(|source| VersioningError::Write { path: file_path, source })?;
    Ok(id)
}

#[cfg(test)]
mod tests;

File: pnpm11/releasing/versioning/src/intents.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import util from 'node:util'

import { PnpmError } from '@pnpm/error'
import { humanId } from 'human-id'
import * as yaml from 'yaml'

export const CHANGES_DIR = '.changeset'

export const BUMP_TYPES = ['none', 'patch', 'minor', 'major'] as const

export type IntentBumpType = typeof BUMP_TYPES[number]

export type ReleaseBumpType = Exclude<IntentBumpType, 'none'>

export interface ChangeIntent {
  id: string
  filePath: string
  releases: Record<string, IntentBumpType>
  summary: string
}

export function parseChangeIntent (content: string, id: string, filePath: string): ChangeIntent {
  const lines = content.replace(/^\uFEFF/, '').split(/\r?\n/)
  const closingIndex = lines[0]?.trim() === '---'
    ? lines.findIndex((line, index) => index > 0 && line.trim() === '---')
    : -1
  if (closingIndex === -1) {
    throw new PnpmError('INVALID_CHANGE_INTENT', `Change intent file ${filePath} has no YAML frontmatter`)
  }

  let frontmatter: unknown
  try {
    frontmatter = yaml.parse(lines.slice(1, closingIndex).join('\n')) ?? {}
  } catch (err: unknown) {
    throw new PnpmError('INVALID_CHANGE_INTENT', `Change intent file ${filePath} has invalid YAML frontmatter: ${util.types.isNativeError(err) ? err.message : String(err)}`)
  }

  if (typeof frontmatter !== 'object' || frontmatter === null || Array.isArray(frontmatter)) {
    throw new PnpmError('INVALID_CHANGE_INTENT', `Change intent file ${filePath} frontmatter must be a mapping of package names to bump types`)
  }

  const releases: Record<string, IntentBumpType> = {}
  for (const [pkgName, bumpType] of Object.entries(frontmatter)) {
    if (typeof bumpType !== 'string' || !(BUMP_TYPES as readonly string[]).includes(bumpType)) {
      throw new PnpmError('INVALID_CHANGE_INTENT', `Change intent file ${filePath} declares an invalid bump type for ${pkgName}: ${String(bumpType)}. Expected one of ${BUMP_TYPES.join(', ')}`)
    }
    releases[pkgName] = bumpType as IntentBumpType
  }

  return {
    id,
    filePath,
    releases,
    summary: lines.slice(closingIndex + 1).join('\n').trim(),
  }
}

export async function readChangeIntents (workspaceDir: string): Promise<ChangeIntent[]> {
  const changesDir = path.join(workspaceDir, CHANGES_DIR)
  let fileNames: string[]
  try {
    fileNames = await fs.readdir(changesDir)
  } catch (err: unknown) {
    if (util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT') {
      return []
    }
    throw err
  }

  const intentFileNames = fileNames
    .filter((fileName) => fileName.endsWith('.md') && fileName.toLowerCase() !== 'readme.md')
    .sort()

  return Promise.all(intentFileNames.map(async (fileName) => {
    const filePath = path.join(changesDir, fileName)
    const content = await fs.readFile(filePath, 'utf8')
    return parseChangeIntent(content, fileName.slice(0, -'.md'.length), filePath)
  }))
}

export interface WriteChangeIntentOptions {
  releases: Record<string, IntentBumpType>
  summary: string
}

export async function writeChangeIntent (workspaceDir: string, opts: WriteChangeIntentOptions): Promise<string> {
  const changesDir = path.join(workspaceDir, CHANGES_DIR)
  await fs.mkdir(changesDir, { recursive: true })

  const existing = new Set(await fs.readdir(changesDir))
  let id = humanId({ separator: '-', capitalize: false })
  while (existing.has(`${id}.md`)) {
    id = humanId({ separator: '-', capitalize: false })
  }

  const frontmatterLines = Object.entries(opts.releases)
    .map(([pkgName, bumpType]) => `${JSON.stringify(pkgName)}: ${bumpType}`)
  const content = `---\n${frontmatterLines.join('\n')}\n---\n\n${opts.summary.trim()}\n`
  await fs.writeFile(path.join(changesDir, `${id}.md`), content, 'utf8')
  return id
}
